<launch>
  <arg name="mav_name" default="$(optenv MAV_NAME)"/>
  <arg name="mav_type" default="$(optenv MAV_TYPE)"/>
  <arg name="mass" default="$(optenv MAV_MASS)"/>

  <arg name="use_vicon" default="false"/>
  <arg name="odom_topic" value="odom" if="$(arg use_vicon)"/>
  <arg name="odom_topic" value="quadrotor_ukf/control_odom" unless="$(arg use_vicon)"/>
  <arg name="odom_topic_throttled" value="quadrotor_ukf/control_odom_throttled"/>
  <arg name="odom_throttle_rate" default="30"/>

  <group ns="$(arg mav_name)">

    <param name="mass" value="$(arg mass)"/>

    <node pkg="nodelet"
      type="nodelet"
      name="controller_nodelet"
      args="manager"
      output="screen">
     <rosparam file="$(find tag_swarm)/config/mav_manager_params.yaml" />
   </node>

    <node pkg="nodelet"
      type="nodelet"
      name="trackers_manager"
      args="load trackers_manager/TrackersManager controller_nodelet"
      output="screen">
      <rosparam file="$(find tag_swarm)/config/trackers.yaml"/>
      <rosparam file="$(find tag_swarm)/config/tracker_params.yaml"/>
      <remap from="~odom" to="$(arg odom_topic)"/>
      <remap from="~cmd" to="position_cmd"/>
    </node>

    <node pkg="nodelet"
      type="nodelet"
      args="load so3_control/SO3ControlNodelet controller_nodelet"
      name="so3_control"
      required="true"
      output="screen">
      <rosparam file="$(find tag_swarm)/config/$(arg mav_type)_gains.yaml"/>
      <param name="mass" value="$(arg mass)"/>
      <param name="use_external_yaw" value="false"/>
      <remap from="~odom" to="$(arg odom_topic)"/>
      <remap from="~position_cmd" to="position_cmd"/>
      <remap from="~so3_cmd" to="so3_cmd"/>
      <remap from="~motors" to="motors"/>
    </node>

    <node pkg="nodelet"
      type="nodelet"
      args="load qualcomm_interface/SO3CmdToQualcomm controller_nodelet"
      name="so3cmd_to_qualcomm_nodelet"
      required="true"
      clear_params="true"
      output="screen">
      <remap from="~odom" to="$(arg odom_topic)"/>
      <remap from="~so3_cmd" to="so3_cmd"/>
    </node>

    <node pkg="mav_manager"
      type="mav_services"
      name="mav_services"
      output="screen">
      <rosparam file="$(find tag_swarm)/config/mav_manager_params.yaml"/>
      <remap from="odom" to="$(arg odom_topic)"/>
    </node>

    <node name="odom_throttler"
      type="throttle"
      pkg="topic_tools"
      args="messages $(arg odom_topic) $(arg odom_throttle_rate) $(arg odom_topic_throttled)"/>

  </group>
</launch>
